### ISSUE TITLE
tf.contrib.cudnn_rnn.CudnnGRU does not work with input_mode='skip_input' in TF1.5

### ISSUE THREAD

**Issue by standy66** [tensorflow/tensorflow#16042]
### System information
- **OS Platform and Distribution**: Linux Ubuntu 16.04
- **TensorFlow installed from**: `pip install tensorflow-gpu==1.5.0rc0`
- **TensorFlow version**: v1.3.0-rc1-6090-g622487f 1.5.0-rc0
- **Python version**:  3.5
- **CUDA/cuDNN version**: CUDA 9.0, CuDNN 7.0.5
- **GPU model and memory**: GTX 1080 8GB

### Describe the problem
We are running tf.contrib.cudnn_rnn.CudnnGRU in our speech recognition setup with input_mode='skip_input' and it crashes the whole process. Here is the assertion error that we are getting:

`Check failed: size == params_input[i].NumElements() Params size mismatch. Expected 0, got 10000`

Looks like it's crashing here: [tensorflow/contrib/cudnn_rnn/kernels/cudnn_rnn_ops.cc:440](https://github.com/tensorflow/tensorflow/blob/r1.5/tensorflow/contrib/cudnn_rnn/kernels/cudnn_rnn_ops.cc#L440). It does not crash if input_mode is 'linear_input'.

Here is the minimal example to reproduce:

### Source code / logs
```
import tensorflow as tf

layer = tf.contrib.cudnn_rnn.CudnnGRU(num_layers=1, num_units=100,
                             input_mode='skip_input', direction='bidirectional')
# (time, batch_size, num_inputs)
x = tf.random_normal((100, 16, 100))
y = layer(x)

with tf.Session() as sess:
	sess.run(tf.global_variables_initializer())
	print(sess.run(y))

```

Logs:

```
2018-01-11 17:10:44.858413: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2018-01-11 17:10:45.067854: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1105] Found device 0 with properties: 
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.8095
pciBusID: 0000:03:00.0
totalMemory: 7.92GiB freeMemory: 7.80GiB
2018-01-11 17:10:45.274222: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1105] Found device 1 with properties: 
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.8095
pciBusID: 0000:04:00.0
totalMemory: 7.92GiB freeMemory: 7.80GiB
2018-01-11 17:10:45.274776: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Device peer to peer matrix
2018-01-11 17:10:45.274804: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1126] DMA: 0 1 
2018-01-11 17:10:45.274812: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1136] 0:   Y Y 
2018-01-11 17:10:45.274817: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1136] 1:   Y Y 
2018-01-11 17:10:45.274826: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1195] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 1080, pci bus id: 0000:03:00.0, compute capability: 6.1)
2018-01-11 17:10:45.274832: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1195] Creating TensorFlow device (/device:GPU:1) -> (device: 1, name: GeForce GTX 1080, pci bus id: 0000:04:00.0, compute capability: 6.1)
2018-01-11 17:10:46.240862: F tensorflow/contrib/cudnn_rnn/kernels/cudnn_rnn_ops.cc:440] Check failed: size == params_input[i].NumElements() Params size mismatch. Expected 0, got 10000
Aborted (core dumped)
```

### Thoughts

I think the problem is in python wrapper code located at [contrib/cudnn_rnn/python/layers/cudnn_rnn.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cudnn_rnn/python/layers/cudnn_rnn.py).

Both [`_canonical_weight_shape(self, layer)`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cudnn_rnn/python/layers/cudnn_rnn.py#L417) and [`_canonical_bias_shape(self, layer)`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cudnn_rnn/python/layers/cudnn_rnn.py#L443) don't handle the case when `input_mode='skip_input'`. The wrapper code doesn't even check if `input_size == num_units` when `input_mode='skip_input'` as it should! The issue seems to be easy to fix, but I may be mistaken.

Here is an example that shows that both 'skip_input' and 'linear_input' get the same set of canonical parameter shapes:
```
CudnnGRU = lambda **kwargs: tf.contrib.cudnn_rnn.CudnnGRU(num_layers=1, num_units=7, direction='bidirectional', **kwargs)

layer = CudnnGRU(input_mode='skip_input')
layer.build((200, 16, 5))
print("skip_input weights", layer.canonical_weight_shapes)
print("skip_input biases", layer.canonical_bias_shapes)

layer = CudnnGRU(input_mode='linear_input')
layer.build((200, 16, 5))
print("linear_input_weights", layer.canonical_weight_shapes)
print("linear_input_biases", layer.canonical_bias_shapes)
```

Example output:
```
skip_input weights [(7, 5), (7, 5), (7, 5), (7, 7), (7, 7), (7, 7), (7, 5), (7, 5), (7, 5), (7, 7), (7, 7), (7, 7)]
skip_input biases [[7], [7], [7], [7], [7], [7], [7], [7], [7], [7], [7], [7]]
linear_input_weights [(7, 5), (7, 5), (7, 5), (7, 7), (7, 7), (7, 7), (7, 5), (7, 5), (7, 5), (7, 7), (7, 7), (7, 7)]
linear_input_biases [[7], [7], [7], [7], [7], [7], [7], [7], [7], [7], [7], [7]]
```

### Workaround

In our setup we decided that we will keep allocating opaque_params buffer manually with first querying its size with 
```
# Note: this code may require tf==1.4 to run.
layer = tf.contrib.cudnn_rnn.CudnnGRU(1, num_units=800, input_size=800,
                                      input_mode='linear_input',
                                      direction='bidirectional')
with tf.Session() as sess:
    OPAQUE_BUFFER_SIZE = sess.run(layer.params_size())
    print(OPAQUE_BUFFER_SIZE)
```

But this is really troublesome because it requires running the intermediate graph to get OPAQUE_BUFFER_SIZE before building the main graph.

**Comment by protoget** [comment]
Hi standy66@, thanks for reporting the issue. Yes only linear_input input_mode is handled for the saveable. The reason is we intend to make sure cudnn-trained checkpoint can be used by regular TF RNNs. Skipping the input isn't what regular TF RNNs do.

I wonder if it would be a problem if we remove the input_mode option from the cudnn_rnn layers API? If one uses cudnn_rnn_ops, they will still be functional, only that users will need to manage saving/restoring variables in canonical format themselves.

**Comment by standy66** [comment]
Hello, @protoget, thanks for quick reply.

First, let me introduce a strong case where skip_input is useful. We are following [DeepSpeech 2](https://arxiv.org/pdf/1512.02595.pdf) paper from Baidu Research. They are using batch normalization for recurrent neural networks similarly to [Laurent et al., 2015](https://arxiv.org/pdf/1510.01378.pdf). Here is an excerpt from the paper:


![](http://oi66.tinypic.com/sw8h9i.jpg)


And in GRUs and LSTMs, they suggest replacing `Wx + b` parts in equiations with `BN(Wx)`. One of the authors of original DeepSpeech 2 paper confirmed that on [Reddit](https://www.reddit.com/r/MachineLearning/comments/4082yn/rnn_batch_normalization_for_grulstm/cyw4ezu/). This is easily achievable by stacking dense layer, batch normalization and recurrent layer with `input_mode=skip_input`.


As for the answer for your question, yes, I think removing input_mode from cudnn_rnn layers API is ok, but then `cudnn_rnn_ops` should be accessibe in documentation (as far as I understand now it's not the case with TF 1.5), maybe with this particular case as an example of where `skip_input` is needed.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by thomasquintana** [comment]
Any update on this issue?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 79 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 94 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 109 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 124 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 139 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 154 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 169 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 184 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 199 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 214 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 229 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 244 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 259 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ymodak** [comment]
TF is moving out of contrib module. Try using ```tf.keras.layers.CuDNNGRU``` instead.
Closing this issue due to staleness. Please check with the latest version of TensorFlow. Feel free to reopen if issue still persists. Thanks!

**Comment by standy66** [comment]
`tf.keras.layers.CuDNNGRU` does not support `skip_input`. It's a shame tensorflow stepped down from providing low-level CuDNN-like interface.
